apps/api/src/scraper/scrapeURL/lib/fetch.ts

```diff
@@ -0,0 +1,144 @@
+import { Logger } from "winston";
+import { z, ZodError } from "zod";
+import { v4 as uuid } from "uuid";
+import * as Sentry from "@sentry/node";
+import FormData from "form-data";
+
+export type RobustFetchParams<Schema extends z.Schema<any>> = {
+    url: string;
+    logger: Logger,
+    method: "GET" | "POST" | "DELETE" | "PUT";
+    body?: any;
+    headers?: Record<string, string>;
+    schema?: Schema;
+    dontParseResponse?: boolean;
+    ignoreResponse?: boolean;
+    ignoreFailure?: boolean;
+    requestId?: string;
+    tryCount?: number;
+};
+
+export async function robustFetch<Schema extends z.Schema<any>, Output = z.infer<Schema>>({
+    url,
+    logger,
+    method = "GET",
+    body,
+    headers,
+    schema,
+    ignoreResponse = false,
+    ignoreFailure = false,
+    requestId = uuid(),
+    tryCount = 1,
+}: RobustFetchParams<Schema>): Promise<Output> {
+    const params = { url, logger, method, body, headers, schema, ignoreResponse, ignoreFailure, tryCount };
+
+    logger.debug("Sending request...", { params, requestId });
+
+    let request: Response;
+    try {
+        request = await fetch(url, {
+            method,
+            headers: {
+                ...(body instanceof FormData
+                ? body.getHeaders()
+                : body !== undefined ? ({
+                    "Content-Type": "application/json",
+                }) : {}),
+                ...(headers !== undefined ? headers : {}),
+            },
+            ...(body instanceof FormData ? ({
+                body: body.getBuffer(),
+            }) : body !== undefined ? ({
+                body: JSON.stringify(body),
+            }) : {}),
+        });
+    } catch (error) {
+        if (!ignoreFailure) {
+            Sentry.captureException(error);
+            if (tryCount > 1) {
+                logger.debug("Request failed, trying " + (tryCount - 1) + " more times", { params, error, requestId });
+                return await robustFetch({
+                    ...params,
+                    requestId,
+                    tryCount: tryCount - 1,
+                });
+            } else {
+                logger.debug("Request failed", { params, error, requestId });
+                throw new Error("Request failed", {
+                    cause: {
+                        params, requestId, error,
+                    },
+                });
+            }
+           
+        } else {
+            return null as Output;
+        }
+    }
+
+    if (ignoreResponse === true) {
+        return null as Output;
+    }
+
+    const response = {
+        status: request.status,
+        headers: request.headers,
+        body: await request.text(), // NOTE: can this throw an exception?
+    };
+
+    if (request.status >= 300) {
```

Comment:
again dont think this is a valid way to define success

Reply:
It is for what robustFetch is used for, which is making rudimentary API calls. This is a non-issue.

Reply:
what is robustFetch?

Reply:
Replacing `axios`, used to make calls to `fire-engine`, `scrapingbee`, the playwright microservice, etc.

Reply:
oh ok, if its internal only we are good. Check the other comment about the same issue, I think that might be for non internal stuff